### PR TITLE
JGoogleEmbedMaps::addEventHandler() to map

### PR DESCRIPTION
**Pull request by itbra**
I found that when creating a map it is not considered to bind events to it. Isn't it a tyical scenario to click on a map and not only zoom in but also trigger some action? 

The [Google Maps API](https://developers.google.com/maps/documentation/javascript/reference) presents as an example to create the map and immediately create the desired event which expects to receive the map created as reference to bind to. I added the functions required for _map event_ handler creation and deletion _([find a PR for marker events here](https://github.com/joomla/joomla-cms/pull/6381))_. Also, the function creating the script was extended to consider registered map event handlers.

I tested this functionality by taking a view template and paste the following:

```
// Target element
echo '<h1>JGoogleEmbedMaps Test</h1>';
echo '<div class="row">';
echo    '<div class="col col-xs-12">';
echo        '<div id="map" style="height:500px; background:gainsboro; border:gray">LOADING ...</div>';
echo    '</div>';
echo    '<div class="col col-xs-12">';
echo        '<h5>Monitor</h5>';
echo        '<p id="monitor" style="padding:1em 0.5em; border:1px solid gainsboro; border-radius:3px"></p>';
echo    '</div>';
echo '</div>';

// Config
$options = new JRegistry(array(
    'key'         => '',
    'mapid'       => 'map',
    'mapclass'    => 'mapclass'
));

// Get instance
$google = new JGoogle($options);
$map = $google->embed('Maps');

// Define load type
$map->setAutoload('jquery');

$map->setMapType('HYBRID');
$map->setCenter(array('50','10'), null);
$map->setZoom('3');

// Marker configuration
$markerOptions = array(
    'clickable' => true,
    'draggable' => false,
    'opacity'   => 0.75,
    'optimized' => true,
    'visible'   => true
);

// Add events handlers
$map->addEventHandler('bounds_changed',     'function(){ document.getElementById("monitor").innerHTML = "map bounds_changed event"; }');
$map->addEventHandler('center_changed',     'function(){ document.getElementById("monitor").innerHTML = "map center_changed event"; }');
$map->addEventHandler('click',              'function(){ alert("map click event"); }');
$map->addEventHandler('dblclick',           'function(){ alert("map dblclick event"); }');
$map->addEventHandler('drag',               'function(){ document.getElementById("monitor").innerHTML = "map drag event"; }');
$map->addEventHandler('dragend',            'function(){ document.getElementById("monitor").innerHTML = "map dragend event"; }');
$map->addEventHandler('dragstart',          'function(){ document.getElementById("monitor").innerHTML = "map dragstart event"; }');
$map->addEventHandler('heading_changed',    'function(){ document.getElementById("monitor").innerHTML = "map heading_changed event"; }');
$map->addEventHandler('idle',               'function(){ document.getElementById("monitor").innerHTML = "map idle event"; }');
$map->addEventHandler('maptypeid_changed',  'function(){ document.getElementById("monitor").innerHTML = "map maptypeid_changed event"; }');
$map->addEventHandler('mousemove',          'function(){ document.getElementById("monitor").innerHTML = "map mousemove event"; }');
$map->addEventHandler('mouseout',           'function(){ document.getElementById("monitor").innerHTML = "map mouseout event"; }');
$map->addEventHandler('mouseover',          'function(){ document.getElementById("monitor").innerHTML = "map mouseover event"; }');
$map->addEventHandler('projection_changed', 'function(){ document.getElementById("monitor").innerHTML = "map projection_changed event"; }');
$map->addEventHandler('resize',             'function(){ document.getElementById("monitor").innerHTML = "map resize event"; }');
$map->addEventHandler('rightclick',         'function(){ document.getElementById("monitor").innerHTML = "map rightclick event"; }');
$map->addEventHandler('tilesloaded',        'function(){ document.getElementById("monitor").innerHTML = "map tilesloaded event"; }');
$map->addEventHandler('tilt_changed',       'function(){ document.getElementById("monitor").innerHTML = "map tilt_changed event"; }');
$map->addEventHandler('zoom_changed',       'function(){ document.getElementById("monitor").innerHTML = "map zoom_changed event"; }');

// Prepare custom script(s)
// initialization callback
$callback =
<<<ENDSCRIPT
window.gm = map;   // Will execute after initialisation to add reference to the created map as a global.
(function($, gm, element) {
    $('#monitor').append('<em>Calling back</em>');
})(jQuery.noConflict(), window.google || {}, document.getElementById({$options->get('mapid')}));
ENDSCRIPT;

// Register custom script(s)
$map->setAdditionalJavascript($callback);

$map->echoHeader();
$map->echoBody();
```

What do you think about this?

If somebody experienced in writing Unit tests might help out to write 3 test for the 3 new functions i would really appreciate your support as i have no experience in that.

### CODE REVIEW

File: libraries/joomla/google/embed/maps.php [hunk 1]
```diff
@@ -318,6 +318,71 @@ public function setCenter($location, $title = true, $markeroptions = array())
 	}
 
 	/**
+	 * Method to add and event handler to the map.
```
Comment (Erftralle) [inline libraries/joomla/google/embed/maps.php]: Only a misspelling, **an event**

File: libraries/joomla/google/embed/maps.php [hunk 2]
```diff
@@ -318,6 +318,71 @@ public function setCenter($location, $title = true, $markeroptions = array())
...
+
+		$events[$type] = $function;
+
+		$this->setOption('events', $events);
+
+		return $this;
+	}
+
+	/**
+	 * Method to remove and event handler from the map
```
Comment (Erftralle) [inline libraries/joomla/google/embed/maps.php]: Only a misspelling, an event

File: libraries/joomla/google/embed/maps.php [hunk 3]
```diff
@@ -318,6 +318,71 @@ public function setCenter($location, $title = true, $markeroptions = array())
 	}
 
 	/**
+	 * Method to add and event handler to the map.
+	 * Event handlers must be passed in without being wrapped by a `function(){}` declaration.
```
Comment (Erftralle) [inline libraries/joomla/google/embed/maps.php]: Why ```without being wrapped```, in you sample code it is wrapped or am I missunderstanding something?

Comment (itbra) [inline libraries/joomla/google/embed/maps.php]: Absolutely! My bad. In fact this was how i had it before i decided to allow fully qualified function definitions. I simply just forgot to change it there.

**Comment by Erftralle** [comment]
Thanks for this usefull addition.

@test: Works fine for me.

**Comment by Rivaner** [comment]
I have tested this item :white_check_mark: successfully on 72107d023295995d3d57062efdc4d29cf96e44fa<br /><br />Patch works!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6472">issues.joomla.org/joomla-cms/6472</a>.</sub>

**Comment by farrelljsec** [comment]
I have tested this item :white_check_mark: successfully on 72107d023295995d3d57062efdc4d29cf96e44fa<br /><br />Patch works!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6472">issues.joomla.org/joomla-cms/6472</a>.</sub>

**Comment by zero-24** [comment]
RTC. Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6472">issues.joomla.org/joomla-cms/6472</a>.</sub>

**Comment by roland-d** [comment]
Thanks everybody. Merged into 3.5-dev with commit  fbd77ac

**Pull request closed without merging**